mypy/stubutil.py

```diff
@@ -106,32 +123,146 @@ def write_header(file: IO[str], module_name: Optional[str] = None,
+    INIT = 1
+    FUNCTION_NAME = 2
+    ARGUMENT_LIST = 3
+    ARGUMENT_TYPE = 4
+    ARGUMENT_DEFAULT = 5
+    RETURN_VALUE = 6
+    OPEN_BRACKET = 7
+
+
+def infer_sig_from_docstring(docstr: str, name: str) -> Optional[List[TypedFunctionSig]]:
     if not docstr:
         return None
-    docstr = docstr.lstrip()
-    # look for function signature, which is any string of the format
-    # <function_name>(<signature>) -> <return type>
-    # or perhaps without the return type
-
-    # in the signature, we allow the following characters:
-    # colon/equal: to match default values, like "a: int=1"
-    # comma/space/brackets: for type hints like "a: Tuple[int, float]"
-    # dot: for classes annotating using full path, like "a: foo.bar.baz"
-    # to capture return type,
-    sig_str = r'\([a-zA-Z0-9_=:, \[\]\.]*\)'
-    sig_match = r'%s(%s)' % (name, sig_str)
-    # first, try to capture return type; we just match until end of line
-    m = re.match(sig_match + ' -> ([a-zA-Z].*)$', docstr, re.MULTILINE)
-    if m:
-        # strip potential white spaces at the right of return type
-        return m.group(1), m.group(2).rstrip()
-
-    # try to not match return type
-    m = re.match(sig_match, docstr)
-    if m:
-        return m.group(1), 'Any'
-    return None
+
+    state = [State.INIT, ]
+    accumulator = ""
+    arg_type = None
+    arg_name = ""
+    arg_default = None
+    ret_type = "Any"
+    found = False
+    args = []  # type: List[TypedArgSig]
+    signatures = []  # type: List[TypedFunctionSig]
+    try:
+        for token in tokenize.tokenize(io.BytesIO(docstr.encode('utf-8')).readline):
+            if token.type == tokenize.NAME and token.string == name and state[-1] == State.INIT:
+                state.append(State.FUNCTION_NAME)
+
+            elif token.type == tokenize.OP and token.string == '(' and state[-1] == \
+                    State.FUNCTION_NAME:
+                state.pop()
+                accumulator = ""
+                found = True
+                state.append(State.ARGUMENT_LIST)
+
+            elif state[-1] == State.FUNCTION_NAME:
+                # reset state, function name not followed by '('
+                state.pop()
+
+            elif token.type == tokenize.OP and token.string in ('[', '(', '{') and \
+                    state[-1] != State.INIT:
+                accumulator += token.string
+                state.append(State.OPEN_BRACKET)
+
+            elif token.type == tokenize.OP and token.string in (']', ')', '}') and \
+                    state[-1] == State.OPEN_BRACKET:
+                accumulator += token.string
+                state.pop()
+
+            elif token.type == tokenize.OP and token.string == ':' and \
+                    state[-1] == State.ARGUMENT_LIST:
+                arg_name = accumulator
+                accumulator = ""
+                state.append(State.ARGUMENT_TYPE)
+
+            elif token.type == tokenize.OP and token.string == '=' and state[-1] in (
```

Comment:
I'd rather do the same reformatting here (and again below).

Reply:
done